api/tools/src/main/java/org/pmiops/workbench/tools/elastic/BigQueryOmopIndexer.java

```diff
@@ -0,0 +1,325 @@
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.Storage.BlobListOption;
+import com.google.cloud.storage.StorageOptions;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Iterables;
+import com.google.common.io.Resources;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.nio.charset.Charset;
+import java.time.Instant;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.Spliterators;
+import java.util.logging.Logger;
+import java.util.stream.StreamSupport;
+import org.apache.commons.cli.CommandLine;
+import org.apache.commons.cli.DefaultParser;
+import org.apache.commons.cli.Option;
+import org.apache.commons.cli.Options;
+import org.apache.http.HttpHost;
+import org.elasticsearch.ElasticsearchStatusException;
+import org.elasticsearch.action.admin.indices.create.CreateIndexRequest;
+import org.elasticsearch.action.admin.indices.delete.DeleteIndexRequest;
+import org.elasticsearch.action.admin.indices.get.GetIndexRequest;
+import org.elasticsearch.action.bulk.BulkItemResponse;
+import org.elasticsearch.action.bulk.BulkRequest;
+import org.elasticsearch.action.bulk.BulkResponse;
+import org.elasticsearch.action.index.IndexRequest;
+import org.elasticsearch.client.RequestOptions;
+import org.elasticsearch.client.RestClient;
+import org.elasticsearch.client.RestHighLevelClient;
+import org.springframework.boot.CommandLineRunner;
+import org.springframework.boot.builder.SpringApplicationBuilder;
+import org.springframework.context.annotation.Bean;
+import org.threeten.bp.Duration;
+
+/**
+ * Command-line tool for indexing an OMOP BigQuery dataset into Elasticsearch. Executes the
+ * following pipeline:
+ *
+ * - Create (and optionally replace) an existing index for this CDR.
+ * - Query BigQuery to produce JSON documents: the output of the query is the exact format we define
+ *   for Elasticsearch (using a repeated STRUCT for repeated nested documents).
+ * - [Optional] Export these results to a temporary table, and export that to GCS. Slower, but
+ *   necessary for larger datasets (e.g. the entire 1m participant Synthetic CDR).
+ * - Bulk ingest the JSON documents into Elasticsearch (either inserts or updates).
+ *
+ * This indexer can optionally be run on a subsample of participants in the CDR for faster test
+ * iterations and local development. The full index will be 10-100GB, so it is not advisable to run
+ * this locally against the entire synthetic CDR.
+ */
+public final class BigQueryOmopIndexer {
+
+  private static Option queryProjectIdOpt = Option.builder()
+      .longOpt("query-project-id")
+      .desc("GCP Cloud Project ID to run the extraction query in. Does not necessarily need " +
+          "to be the same as the CDR dataset's BigQuery project.")
+      .required()
+      .hasArg()
+      .build();
+  private static Option esBaseUrlOpt = Option.builder()
```

Comment:
I think I prefer the bulider form since some of these need slightly different modifiers. I did a few iterations of this and I landed on this as an "OK" solution here. The thing I'd lose if I used that approach (if I'm understanding correctly), is the ability to reference the options throughout the code. I initially was retyping the flag strings, which was error-prone. Moving those strings to constants feels like higher boilerplate than this approach to me, though that might be wrong.